### PR TITLE
AEM Article Database

### PR DESCRIPTION
**Pull request by gyasistory**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -115,19 +115,17 @@ dependencies {
...
     implementation project(':library:sync')
-
     implementation project(':ui:account')
     implementation project(':ui:base')
     implementation project(':ui:base-app')
     implementation project(':ui:everystudent')
     implementation project(':ui:shortcuts')
+    implementation project(':ui:articles-aem-renderer')
     implementation project(':ui:tract-renderer')
-
```
Comment (frett) [inline app/build.gradle]: guessing that you used the Android Studio dependency editor, it tends to strip out extra whitespace :)

I keep the whitespace gaps to help group related dependencies together for organization and ease of seeing what's included.

I always alpha sort dependencies within each section

File: build.gradle [hunk 2]
```diff
@@ -68,7 +68,9 @@ ext.deps = [
         retrofit2            : '2.4.0',
         snowplow             : '0.7.0',
         thekey               : '2.0.0-SNAPSHOT',
-        timber               : '4.7.0'
+        timber               : '4.7.0',
+        roomVersion          : '1.0.0',
+        archLifecycleVersion : '1.1.0'
```
Comment (frett) [inline build.gradle]: as I think you noticed elsewhere, I already had this version defined using `androidLifecycle`. I'm fine with using either name, but we should only use one and not both to make sure we can keep versions correctly in sync.

File: settings.gradle [hunk 3]
```diff
@@ -1,3 +1,4 @@
+include ':articles-aem-renderer'
```
Comment (frett) [inline settings.gradle]: I don't think this module actually exists, so this line can be removed :)

File: ui/articles-aem-renderer/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+/build
```
Comment (frett) [inline ui/articles-aem-renderer/.gitignore]: the root `.gitignore` already takes care of this, so there isn't a need to include the additional `.gitignore` file

File: ui/articles-aem-renderer/build.gradle [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        vectorDrawables.useSupportLibrary = true
+
+        testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
+
+        javaCompileOptions {
+            annotationProcessorOptions {
+                arguments = [eventBusIndex: 'org.cru.godtools.tract.TractEventBusIndex']
+            }
+        }
+        multiDexEnabled true
```
Comment (frett) [inline ui/articles-aem-renderer/build.gradle]: you shouldn't be enabling multi-dex for a library dependency, that should be defined at the application level.

File: ui/articles-aem-renderer/build.gradle [hunk 6]
```diff
@@ -0,0 +1,44 @@
+apply plugin: 'com.jakewharton.butterknife'
+
+android {
+    defaultConfig {
+        vectorDrawables.useSupportLibrary = true
```
Comment (frett) [inline ui/articles-aem-renderer/build.gradle]: I believe this is already taken care of in the root build.gradle file

File: ui/articles-aem-renderer/src/main/java/org/godtools/uiarticles_aem_renderer/db/ArticleDao.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package org.godtools.uiarticles_aem_renderer.db;
```
[the remark this answers is not in the document]
Comment (gyasistory) [reply ui/articles-aem-renderer/src/main/java/org/godtools/uiarticles_aem_renderer/db/ArticleDao.java]: done

**Comment by frett** [comment]
As you continue to work on this I'll keep dropping comments in here for small style tweaks and other things :)

File: ui/articles-aem-renderer/build.gradle (hunk 6 above)
Comment (gyasistory) [reply ui/articles-aem-renderer/build.gradle]: This was in the Trac-Render gradle file so that is why I added it.  I can remove it if you think it is unnecessary.

File: ui/articles-aem-renderer/build.gradle (hunk 5 above)
Comment (gyasistory) [reply ui/articles-aem-renderer/build.gradle]: I was getting an error about it being only a single dex this is why I added it.

File: ui/articles-aem-renderer/.gitignore (hunk 4 above)
Comment (gyasistory) [reply ui/articles-aem-renderer/.gitignore]: The wizard added it I removed it.

File: ui/articles-aem-renderer/build.gradle (hunk 6 above)
Comment (frett) [reply ui/articles-aem-renderer/build.gradle]: let's remove it for the moment until we actually get to the point of adding views and vector images to this module

File: ui/articles-aem-renderer/build.gradle (hunk 5 above)
Comment (frett) [reply ui/articles-aem-renderer/build.gradle]: hmm, in theory that should be handled at the app level only (at least based on my understanding of it)

File: settings.gradle (hunk 3 above)
Comment (gyasistory) [reply settings.gradle]: For some reason when I remove it.  it can find my model
godtools____documents_github_godtools-android__-_____settings_gradle__godtools_

File: ui/articles-aem-renderer/build.gradle (hunk 5 above)
Comment (frett) [reply ui/articles-aem-renderer/build.gradle]: I'll do a little bit of research

File: build.gradle (hunk 2 above)
Comment (gyasistory) [reply build.gradle]: ok

File: settings.gradle (hunk 3 above)
Comment (frett) [reply settings.gradle]: hmm, interesting. leave it for now, I'll take a look later

Comment (gyasistory) [reply settings.gradle]: OK I figured it out the .iml was misconfigured.

Comment (frett) [reply settings.gradle]: makes sense

File: ui/articles-aem-renderer/build.gradle (hunk 5 above)
Comment (frett) [reply ui/articles-aem-renderer/build.gradle]: any library project added to the library or ui subdirectories gets this configuration automatically applied: https://github.com/CruGlobal/godtools-android/blob/develop/build.gradle#L108-L136

it saves having to duplicate a bunch of common configuration across all the modules

**Comment by frett** [comment]
we have some android device unit testing already setup (nothing using espresso). see the tract-renderer module for some of the android device tests. we didn't use any of the `useLibrary` lines they mention in their documentation. we include the test framework dependencies as `androidTestImplementation` dependencies to guarantee that the same test framework runs on all versions of android. and that we can have the latest version of the framework.

**Comment by gyasistory** [comment]
We are probably going to have to sit down at some point and compare our dev - environments.  Because I can't run your test.  But I can run mine.

 
godtools____documents_github_godtools-android__-_____ui_tract-renderer_src_test_java_org_cru_godtools_tract_model_imagegravitytest_java__tract-renderer_

godtools____documents_github_godtools-android__-_____ui_articles-aem-renderer_src_androidtest_java_org_godtools_articles_exampleinstrumentedtest_java__articles-aem-renderer_

**Comment by frett** [comment]
Android testing is a bit of a mess. `ImageGravityTest` is a local unit test and `ExampleInstrumentedTest` is an android unit test. there are 2 types of unit tests within Android

**Comment by frett** [comment]
and it appears that the ImageGravityTest doesn't run inside AS, but runs from the command line just fine...

**Comment by gyasistory** [comment]
That would explain it.  I am kind of deep in this test but once I am finish.  You can tell me what direction you would like to go and we can go from there.

**Comment by frett** [comment]
I fixed the merge conflict in `app/build.gradle` after I merged the xml-model refactor

File: app/build.gradle [hunk 8]
```diff
@@ -122,6 +122,7 @@ dependencies {
     implementation project(':ui:base-app')
     implementation project(':ui:everystudent')
     implementation project(':ui:shortcuts')
+    implementation project(':ui:articles-aem-renderer')
```
Comment (frett) [inline app/build.gradle]: alpha sort within the ui module dependencies

File: articles-aem-renderer/src/androidTest/java/org/godtools/articles_aem_renderer/ExampleInstrumentedTest.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+import static org.junit.Assert.*;
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+@RunWith(AndroidJUnit4.class)
+public class ExampleInstrumentedTest {
```
Comment (frett) [inline articles-aem-renderer/src/androidTest/java/org/godtools/articles_aem_renderer/ExampleInstrumentedTest.java]: this test can be removed, it doesn't do anything, guessing it was an autogenerated test

File: articles-aem-renderer/src/test/java/org/godtools/articles_aem_renderer/ExampleUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+public class ExampleUnitTest {
```
Comment (frett) [inline articles-aem-renderer/src/test/java/org/godtools/articles_aem_renderer/ExampleUnitTest.java]: again, autogenerated test that can be removed

File: build.gradle [hunk 11]
```diff
@@ -68,7 +68,8 @@ ext.deps = [
         retrofit2            : '2.4.0',
         snowplow             : '0.7.0',
         thekey               : '2.0.0-SNAPSHOT',
-        timber               : '4.7.0'
+        timber               : '4.7.0',
+        roomVersion          : '1.0.0'
```
Comment (frett) [inline build.gradle]: alpha sort `roomVersion`. makes it easier to maintain longterm :)

File: library/db/build.gradle [hunk 12]
```diff
@@ -8,4 +8,13 @@ dependencies {
...
     implementation "com.crashlytics.sdk.android:crashlytics:${deps.crashlytics}"
+
+    // Room components
+//    implementation "android.arch.persistence.room:runtime:${deps.roomVersion}"
+//    annotationProcessor "android.arch.persistence.room:compiler:${deps.roomVersion}"
+//    androidTestImplementation "android.arch.persistence.room:testing:${deps.roomVersion}"
+//
+//    // Lifecycle components
+//    implementation "android.arch.lifecycle:extensions:${deps.archLifecycleVersion}"
+//    annotationProcessor "android.arch.lifecycle:compiler:${deps.archLifecycleVersion}"
```
Comment (frett) [inline library/db/build.gradle]: unnecessary commented change, can be removed :)

File: ui/articles-aem-renderer/src/androidTest/java/org/godtools/articles/ExampleInstrumentedTest.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+    private ArticleRepository mArticleDao;
+    private AttachmentRepository mAttachmentDao;
+    private ManifestAssociationRepository mAssociationDao;
+    private ArticleRoomDatabase db;
+
+
+    @Before
+    public void createDb() {
+//        Context context = InstrumentationRegistry.getTargetContext();
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/godtools/articles/ExampleInstrumentedTest.java]: did the room db tests just not work out?

File: ui/articles-aem-renderer/src/main/AndroidManifest.xml [hunk 14]
```diff
@@ -0,0 +1,2 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="org.godtools.articles.aem" />
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/AndroidManifest.xml]: small nit, `package="org.cru.godtools.article.aem"`

File: ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ArticleRoomDatabase.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+    private static ArticleRoomDatabase instance;
+
+    /**
+     * Used to get the current Instance of the Room database.
+     *
+     * @param context = the application context
+     * @return = Instance of the current Room Database
+     */
+    static ArticleRoomDatabase getINSTANCE(final Context context) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ArticleRoomDatabase.java]: camel-case method name `getInstance`

File: ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ArticleRoomDatabase.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * This class is used to create the database table.
+ *
+ * @author Gyasi Story
+ */
+@Database(entities = { Article.class, ManifestAssociation.class, Attachment.class }, version = 1)
+public abstract class ArticleRoomDatabase extends RoomDatabase {
+
+    public abstract ArticleDao mArticleDao();
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ArticleRoomDatabase.java]: since these are dao access methods, let's not use the `m` prefix on them

File: ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/model/Article.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    @NonNull
+    @ColumnInfo(name = "content")
+    public String mContent;
+
+    /**
+     * The date the article was created
+     */
+    @NonNull
+    @ColumnInfo(name = "date_created")
+    public String mDateCreated;
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/model/Article.java]: I know you had mentioned data transformers for dates. It would be great if on the POJO this could be a `Date` field, and it be stored as a `long` within the database.

File: ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ArticleRepository.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        mArticleDao = db.mArticleDao();
+        mAllArticles = mArticleDao.getAllArticles();
+    }
+
+    /**
+     * To call a list of all article stored in the database.
+     *
+     * @return = Live Data Collection of Articles
+     */
+    public LiveData<List<Article>> getmAllArticles() {
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ArticleRepository.java]: `getAllArticles()`

guessing this was autogenerated, there is a way to configure AS to ignore the `m` prefix when autogenerating getters/setters

Comment (frett) [reply ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ArticleRepository.java]: it's under Android Studio -> Preferences -> Editor -> Code Style -> Java -> Code Generation.

add "m" to the Field name prefix.
screen shot 2018-08-24 at 11 47 48 am

File: ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ManifestAssociationRepository.java [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.List;
+
+/**
+ * This class is used a connector for the Data Access Object for Manifest Associations
+ */
+public class ManifestAssociationRepository {
+
+    private ManifestAssociationDao mManifestAssociationDao;
+
+    public ManifestAssociationRepository(Application application) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ManifestAssociationRepository.java]: you can actually pass in any `Context` object, it doesn't need to be the Application object

File: ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ManifestAssociationRepository.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+import org.godtools.articles.aem.model.ManifestAssociation;
+
+import java.util.List;
+
+/**
+ * This class is used a connector for the Data Access Object for Manifest Associations
+ */
+public class ManifestAssociationRepository {
+
+    private ManifestAssociationDao mManifestAssociationDao;
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/db/ManifestAssociationRepository.java]: make this final since you are only setting it when creating this object.

File: ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/view_model/ArticleViewModel.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+import org.godtools.articles.aem.model.Article;
+import org.godtools.articles.aem.model.Attachment;
+
+import java.util.List;
+
+public class ArticleViewModel extends AndroidViewModel {
+
+    private ArticleRepository mArticleRepository;
+    private AttachmentRepository mAttchReposistory;
+    private ManifestAssociationRepository mManifestRepository;
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/godtools/articles/aem/view_model/ArticleViewModel.java]: all final since you only set them in the constructor and never change them

File: ui/articles-aem-renderer/src/androidTest/java/org/godtools/articles/ExampleInstrumentedTest.java (hunk 13 above)
Comment (gyasistory) [reply ui/articles-aem-renderer/src/androidTest/java/org/godtools/articles/ExampleInstrumentedTest.java]: I had the DAOs as public when I created the test.  But I wanted to make them Package Protected and Use the Repositories to access the data.  I didn't need to test the Repositories right now because they are just calling the DAO.  Later the repository may contain more logic for interacting with the database.

File: articles-aem-renderer/src/androidTest/java/org/godtools/articles_aem_renderer/ExampleInstrumentedTest.java (hunk 9 above)
Comment (gyasistory) [reply articles-aem-renderer/src/androidTest/java/org/godtools/articles_aem_renderer/ExampleInstrumentedTest.java]: will do.

File: ui/articles-aem-renderer/src/main/res/values/strings.xml [hunk 22]
```diff
@@ -0,0 +1,3 @@
+<resources>
+    <string name="app_name">ui/articles-aem-renderer</string>
+</resources>
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/res/values/strings.xml]: this needs to go away so it doesn't interfere with the app_name set elsewhere :)

File: ui/articles-aem-renderer/src/test/java/org/godtools/articles/ExampleUnitTest.java [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+public class ExampleUnitTest {
+
+    @Test
+    public void addition_isCorrect() {
+        assertEquals(4, 2 + 2);
+    }
+}
```
Comment (frett) [inline ui/articles-aem-renderer/src/test/java/org/godtools/articles/ExampleUnitTest.java]: autogenerated test which can be deleted

File: ui/articles-aem-renderer/src/androidTest/java/org/godtools/articles/ExampleInstrumentedTest.java (hunk 13 above)
Comment (frett) [reply ui/articles-aem-renderer/src/androidTest/java/org/godtools/articles/ExampleInstrumentedTest.java]: you can still use the tests even if the DAOs are package private, just put them in the same package :)

File: app/build.gradle [hunk 24]
```diff
@@ -165,12 +166,12 @@ dependencies {
...
     // fix conflicts in transitive dependencies
     androidTestImplementation "com.android.support:multidex:${deps.multidex}"
-    androidTestImplementation "com.google.code.findbugs:jsr305:2.0.1"
+    androidTestImplementation 'com.google.code.findbugs:jsr305:2.0.1'
 }
 
 // make crashlytics upload task dependent on assembling the apk
 afterEvaluate {
-    crashlyticsUploadDistributionProductionDebug.dependsOn assembleProductionDebug
+//    crashlyticsUploadDistributionProductionDebug.dependsOn assembleProductionDebug
```
Comment (frett) [inline app/build.gradle]: this now conflicts with the develop branch

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+
+import timber.log.Timber;
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+@RunWith(AndroidJUnit4.class)
+public class ArticleDBTest implements LifecycleOwner {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: Test classes should follow the naming scheme `{classBeingTested}Test`. So in this case this should be `ArticleRoomDatabaseTest`

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+        for (int i = 0; i < 12; i++) {
+            Article article = new Article();
+
+            article.mContent = "<p> The Body </>";
+            try {
+                article.mDateCreated = new SimpleDateFormat("E MMM dd yyyy HH:mm:ss zz")
+                        .parse("Fri Jun 08 2018 18:55:00 GMT+0000").getTime();
+                article.mDateUpdated = new SimpleDateFormat("E MMM dd yyyy HH:mm:ss zz")
+                        .parse("Sat May 19 2018 00:23:39 GMT+0000").getTime();
+            } catch (ParseException e) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: within test classes you shouldn't be catching any exceptions, you should let them throw so that the test actually fails if something goes wrong.
The only time you would catch an Exception is if it is expected, and then you should `fail()` within the test body so you know if the Exception wasn't actually thrown when you were expecting it to throw.

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+
+    }
+
+    @After
+    public void closeDb() {
+        db.close();
+    }
+
+    @Test
+    public void useAppContext() {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: test methods should be named for the specific method/functionality they are testing. for example:
- `verifyGetAllArticles()`
- `testInsertArticle()`
- `testDeleteArticle()`

This helps keep tests focused on testing specific functionality and makes it easier to track down when certain functionality changes.

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+            Calendar calendar = Calendar.getInstance();
+            calendar.getTime();
+            calendar.setTimeInMillis(article.mDateCreated);
+            Date date = new Date(article.mDateCreated);
+            date.getYear();
+            date.getDate();
+            date.getHours();
+            date.getTimezoneOffset();
+            article.mTitle = " The title = " + i;
+            mArticleDao.insertArticle(article);
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: you are inserting new articles, but never verifying that they are actually being stored in the database. This test would have limited utility

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+
+        }
+
+    }
+
+
+    private void associateToArticles(List<Article> savedArticles) {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: this functionality is never called

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 30]
```diff
@@ -0,0 +1,145 @@
...
+            // Create Manifest Association
+            ManifestAssociation association = new ManifestAssociation();
+            association.mArticleId = article.mId;
+            association.mManifestId = String.valueOf(article.mId % 3);  // just need to only have three versions
+            association.mManifestName = String.format("Manifest # %s", association.mManifestId);
+            mAssociationDao.insertAssociation(association);
+        }
+    }
+
+    private void getAllSavedArticles() {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: also not called

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+                    Objects.requireNonNull(mAssociationDao
+                            .getArticlesByManifestID("0").getValue()).size() > 0 &&
+                    Objects.requireNonNull(mAssociationDao
+                            .getAssociationByManifestID("1").getValue()).size() > 0);
+        });
+    }
+
+
+    @Override
+    public Lifecycle getLifecycle() {
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: why did you need to instrument this with `LifecycleObserver`?

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java (hunk 26 above)
Comment (frett) [reply ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: for example:
```java
    @Test
    public void verifyMissingService() throws Exception {
        try {
            controller.getTicket(null, ticketParams(ACCESS_TOKEN, null));
            fail();
        } catch (final OAuthApiError e) {
            assertThat(e, matchesApiError(BAD_REQUEST, ERROR_BEARER_INVALID_REQUEST));
        }
    }
```

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java (hunk 31 above)
Comment (gyasistory) [reply ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: I was having trouble testing the Live Dat in the Test I wrote a task with Immutable data and it worked fine.

File: ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java [hunk 32]
```diff
@@ -90,56 +73,59 @@ public void useAppContext() {
...
-            Boolean allAttachmentsSaved = false;
+    @Test
+    public void verifyGetAllArticles() {
 
-            for (Article article : articles) {
-                if (mAttachmentDao.getAttachmentsByArticle(article.mId) == null) {
-                    allAttachmentsSaved = false;
-                    break;
-                }
+        assert (mSavedArticles.size() > 1);
```
Comment (frett) [inline ui/articles-aem-renderer/src/androidTest/java/org/cru/godtools/articles/aem/db/ArticleDBTest.java]: you should use the junit assert methods. The built in java `assert` is ignored on android under certain scenarios. Plus the JUnit assert methods allow you to specify a message with the error thrown and give more description messages.

File: ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/db/ManifestAssociationDao.java [hunk 33]
```diff
@@ -54,4 +54,14 @@
             "WHERE assc.manifest_key = :manifestID")
     LiveData<List<Article>> getArticlesByManifestID(String manifestID);
 
+
+    //region Testable calls (No Live Data)
```
Comment (frett) [inline ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/db/ManifestAssociationDao.java]: there are some annotations that signify to AS restricted scope so that these methods aren't used in incorrect places.

Comment (frett) [reply ui/articles-aem-renderer/src/main/java/org/cru/godtools/articles/aem/db/ManifestAssociationDao.java]: need to look up which ones are appropriate :)

File: ui/articles-aem-renderer/build.gradle [hunk 34]
```diff
@@ -5,22 +5,14 @@ android {
...
-
-    implementation project(':ui:base')
 
     implementation "android.arch.lifecycle:extensions:${deps.androidLifecycle}"
     implementation "android.arch.persistence.room:runtime:${deps.roomVersion}"
     implementation "com.android.support:appcompat-v7:${deps.androidSupport}"
 
     androidTestImplementation "com.android.support.test:runner:${deps.androidTestingSupport}"
     androidTestImplementation "android.arch.persistence.room:testing:${deps.roomVersion}"
+    testImplementation "android.arch.core:core-testing:${deps.androidLifecycle}"
```
Comment (frett) [inline ui/articles-aem-renderer/build.gradle]: should this have the lifecycle version? or is it a separate architecture component that is versioned independently?

Comment (frett) [reply ui/articles-aem-renderer/build.gradle]: also, `testImplementation` targets the unit tests, not android tests. So, this dependency is never used

**Comment by frett** [comment]
@gyasistory looks good! I'm going to make a couple minor changes to the test to illustrate a couple things, then we should merge this
